rfq/negotiator.go

```diff
@@ -541,6 +578,119 @@ func (n *Negotiator) HasAssetSellOffer(assetID *asset.ID,
 	return true
 }
 
+// BuyOffer is a struct that represents an asset buy offer. This data structure
+// describes the maximum amount of an asset that this node is willing to
+// purchase.
+//
+// A buy offer is passive (unlike a buy order), meaning that it does not
+// actively lead to a buy request being sent to a peer. Instead, it is used by
+// the node to selectively accept or reject incoming asset sell quote requests
+// before price is considered.
+type BuyOffer struct {
+	// AssetID represents the identifier of the subject asset.
+	AssetID *asset.ID
+
+	// AssetGroupKey is the public group key of the subject asset.
+	AssetGroupKey *btcec.PublicKey
+
+	// MaxUnits is the maximum amount of the asset which this node is
+	// willing to purchase.
+	MaxUnits uint64
+}
+
+// Validate validates the asset buy offer.
+func (a *BuyOffer) Validate() error {
+	if a.AssetID == nil && a.AssetGroupKey == nil {
+		return fmt.Errorf("asset ID is nil and asset group key is nil")
+	}
+
+	if a.AssetID != nil && a.AssetGroupKey != nil {
+		return fmt.Errorf("asset ID and asset group key are both set")
+	}
+
+	if a.MaxUnits == 0 {
+		return fmt.Errorf("max asset amount is zero")
+	}
+
+	return nil
+}
+
+// UpsertAssetBuyOffer upserts an asset buy offer. If the offer already exists
+// for the given asset, it will be updated.
+func (n *Negotiator) UpsertAssetBuyOffer(offer BuyOffer) error {
+	// Validate the offer.
+	err := offer.Validate()
+	if err != nil {
+		return fmt.Errorf("invalid asset buy offer: %w", err)
+	}
+
+	// Store the offer in the appropriate map.
+	//
+	// If the asset group key is not nil, then we will use it as the key for
+	// the offer. Otherwise, we will use the asset ID as the key.
+	switch {
+	case offer.AssetGroupKey != nil:
+		n.assetGroupBuyOffers.Store(*offer.AssetGroupKey, offer)
```

Comment:
Didn't spot this on the prior PRs, but the group key should be serialized before being used as a map key.

Reply:
Could you please help me understand why the group key should be serialized before being used as a map key. The key type in this instance is `btcec.PublicKey`, and not a pointer.

Reply:
So `btcec.PublicKey` is just two `secp.FieldVal` values:

https://github.com/decred/dcrd/blob/08d8572807872f2b9737f8a118b16c320a04b077/dcrec/secp256k1/pubkey.go#L69

But the same public key can be represented by multiple combinations of `secp.FieldVal` values (as those values can be in different formats, non-normalized, etc). For map keys that are structs, they will be compared field-by-field:

https://go.dev/ref/spec#Comparison_operators

So the internal `x` and `y` values will be compared. Using the serialized compressed key avoids this issue (which has appeared before with pubkeys as map keys).

Reply:
Ok, I think I understand. Essentially, even though two public keys perform identically, their underlying structures may vary. Consequently, comparing these structures directly, especially the secp.FieldVal fields, might result in an unreliable assessment of equality.